File: walletsdk/src/main/java/leaf/prod/walletsdk/manager/OrderDataManager.java
/**
 * Created with IntelliJ IDEA.
 * User: kenshin wangchen@loopring.org
 * Time: 2018-11-19 5:39 PM
 * Cooperation: loopring.org 路印协议基金会
 */
package leaf.prod.walletsdk.manager;

import java.math.BigInteger;
import java.util.HashMap;
import java.util.Map;

import android.content.Context;

import org.web3j.crypto.Credentials;
import org.web3j.crypto.Hash;
import org.web3j.crypto.Sign.SignatureData;
import org.web3j.utils.Numeric;

import leaf.prod.walletsdk.Default;
import leaf.prod.walletsdk.Erc20TransactionManager;
import leaf.prod.walletsdk.R;
import leaf.prod.walletsdk.Transfer;
import leaf.prod.walletsdk.model.RandomWallet;
import leaf.prod.walletsdk.model.common.TradeType;
import leaf.prod.walletsdk.model.order.Erc1400Params;
import leaf.prod.walletsdk.model.order.FeeParams;
import leaf.prod.walletsdk.model.order.OrderParams;
import leaf.prod.walletsdk.model.order.RawOrder;
import leaf.prod.walletsdk.model.response.RelayError;
import leaf.prod.walletsdk.model.response.RelayResponseWrapper;
import leaf.prod.walletsdk.model.sign.BitStream;
import leaf.prod.walletsdk.service.RelayService;
import leaf.prod.walletsdk.util.SignUtils;
import leaf.prod.walletsdk.util.StringUtils;
import leaf.prod.walletsdk.util.WalletUtil;
import lombok.Getter;
import lombok.Setter;
import rx.Observable;
import rx.functions.Func1;
import rx.schedulers.Schedulers;

@Getter
@Setter
public abstract class OrderDataManager {

    protected String owner;

    // token symbol, e.g. weth
    protected String tokenSell;

    // token symbol, e.g. lrc
    protected String tokenBuy;

    // e.g. lrc-weth
    protected String tradePair;

    protected Context context;

    protected GasDataManager gas;

    protected Credentials credentials;

    protected TokenDataManager token;

    protected BalanceDataManager balance;

    protected Map<String, Double> balanceInfo;

    protected RelayService relayService;

    OrderDataManager(Context context) {
        this.context = context;
        this.balanceInfo = new HashMap<>();
        this.relayService = new RelayService();
        this.owner = WalletUtil.getCurrentAddress(context);
        this.gas = GasDataManager.getInstance(context);
        this.token = TokenDataManager.getInstance(context);
        this.balance = BalanceDataManager.getInstance(context);
    }

    public RawOrder constructOrder(Double amountBuy, Double amountSell, Integer validSince, Integer validUntil) {
        RawOrder order = null;
        try {
            String tokenB = token.getTokenBySymbol(tokenBuy).getProtocol();
            String tokenS = token.getTokenBySymbol(tokenSell).getProtocol();
            String tokenFee = token.getTokenBySymbol("LRC").getProtocol();
            String amountB = Numeric.toHexStringWithPrefix(token.getWeiFromDouble(tokenBuy, amountBuy));
            String amountS = Numeric.toHexStringWithPrefix(token.getWeiFromDouble(tokenSell, amountSell));
            String owner = WalletUtil.getCurrentAddress(context);
            RandomWallet randomWallet = WalletUtil.getRandomWallet();
            OrderParams orderParams = OrderParams.builder()
                    .sig("")
                    .broker("")
                    .orderInterceptor("")
                    .allOrNone(false)
                    .validUntil(validUntil)
                    .dualAuthAddr(randomWallet.getAddress())
                    .dualAuthPrivateKey(randomWallet.getPrivateKey())
                    .wallet(PartnerDataManager.getInstance(context).getWalletAddress())
                    .build();
            FeeParams feeParams = FeeParams.builder()
                    .tokenFee(tokenFee)
                    .amountFee(Numeric.toHexStringWithPrefix(BigInteger.ZERO))
                    .tokenRecipient(owner)
                    .tokenSFeePercentage(0)
                    .tokenBFeePercentage(0)
                    .walletSplitPercentage(50)
                    .build();
            Erc1400Params erc1400Params = Erc1400Params.builder().build();
            order = RawOrder.builder()
                    .owner(owner).version(0)
                    .tokenB(tokenB).tokenBuy(tokenBuy)
                    .tokenS(tokenS).tokenSell(tokenSell)
                    .amountB(amountB).amountBuy(amountBuy)
                    .amountS(amountS).amountSell(amountSell)
                    .validSince(validSince)
                    .params(orderParams)
                    .feeParams(feeParams)
                    .erc1400Params(erc1400Params)
                    .build();
        } catch (Exception e) {
            e.printStackTrace();
        }
        return order;
    }

    public RawOrder signOrder(RawOrder order) {
        String hash = encodeOrder(order);
        String sign = getSignature(hash);
        order.setHash(hash);
        order.getParams().setSig(sign);
        return order;
    }

    private String getSignature(String hash) {
        byte[] data = Numeric.hexStringToByteArray(hash);
        SignatureData sigData = SignUtils.genSignMessage(credentials, data);
        BitStream bitStream = new BitStream();
        bitStream.addNumber(BigInteger.ONE, 1, true);
        bitStream.addNumber(BigInteger.valueOf(32 + 32 + 1), 1, true);
        bitStream.addNumber(BigInteger.valueOf((int) sigData.getV()), 1, true);
        bitStream.addRawBytes(sigData.getR(), true);
        bitStream.addRawBytes(sigData.getS(), true);
        return bitStream.getData();
    }

    private String encodeOrder(RawOrder order) {
        BitStream bitStream = new BitStream();
        FeeParams feeParams = order.getFeeParams();
        OrderParams orderParams = order.getParams();
        Erc1400Params erc1400Params = order.getErc1400Params();
        byte[] transferDataBytes = erc1400Params.getTransferDataS().getBytes();
        String transferDataHash = Numeric.toHexString(Hash.sha3(transferDataBytes));
        bitStream.addBytes32(SignUtils.Eip712OrderSchemaHash, true);
        bitStream.addUint(Numeric.toBigInt(order.getAmountS()), true);
        bitStream.addUint(Numeric.toBigInt(order.getAmountB()), true);
        bitStream.addUint(Numeric.toBigInt(feeParams.getAmountFee()), true);
        bitStream.addUint(BigInteger.valueOf(order.getValidSince()), true);
        bitStream.addUint(BigInteger.valueOf(orderParams.getValidUntil()), true);
        bitStream.addAddress(order.getOwner(), 32, true);
        bitStream.addAddress(order.getTokenS(), 32, true);
        bitStream.addAddress(order.getTokenB(), 32, true);
        bitStream.addAddress(orderParams.getDualAuthAddr(), 32, true);
        bitStream.addAddress(orderParams.getBroker(), 32, true);
        bitStream.addAddress(orderParams.getOrderInterceptor(), 32, true);
        bitStream.addAddress(orderParams.getWallet(), 32, true);
        bitStream.addAddress(feeParams.getTokenRecipient(), 32, true);
        bitStream.addAddress(feeParams.getTokenFee(), true);
        bitStream.addUint(BigInteger.valueOf(feeParams.getWalletSplitPercentage()), true);
        bitStream.addUint(BigInteger.valueOf(feeParams.getTokenSFeePercentage()), true);
        bitStream.addUint(BigInteger.valueOf(feeParams.getTokenBFeePercentage()), true);
        bitStream.addBoolean(orderParams.getAllOrNone(), true);
        bitStream.addUint(BigInteger.valueOf(erc1400Params.getTokenStandardS()), true);
        bitStream.addUint(BigInteger.valueOf(erc1400Params.getTokenStandardB()), true);
        bitStream.addUint(BigInteger.valueOf(erc1400Params.getTokenStandardFee()), true);
        bitStream.addBytes32(erc1400Params.getTrancheS(), true);
        bitStream.addBytes32(erc1400Params.getTrancheB(), true);
        bitStream.addBytes32(transferDataHash, true);
        String orderDataHash = Numeric.toHexString(Hash.sha3(bitStream.getBytes()));
        BitStream outerStream = new BitStream();
        outerStream.addHex(StringUtils.toHex(SignUtils.Eip191Header), true);
        outerStream.addBytes32(SignUtils.Eip712DomainHash, true);
        outerStream.addBytes32(orderDataHash, true);
        return Numeric.toHexString(Hash.sha3(outerStream.getBytes()));
    }

    public TradeType getSide(RawOrder order) {
        MarketPriceDataManager instance = MarketPriceDataManager.getInstance(context);
        return instance.getOrderSide(order.getTokenBuy(), order.getTokenSell());
    }

    public Observable<RelayResponseWrapper> handleInfo() {
        Observable<RelayResponseWrapper> result;
        if (needApprove()) {
            result = approve().observeOn(Schedulers.io())
                    .flatMap((Func1<String, Observable<RelayResponseWrapper>>) hash -> {
                        if (!hash.equals("failed")) {
                            return submit();
                        }
                        RelayError failed = RelayError.builder()
                                .message(context.getString(R.string.approve_failed)).build();
                        RelayResponseWrapper<Object> response = RelayResponseWrapper.builder().error(failed).build();
                        return Observable.just(response);
                    });
        } else {
            result = submit();
        }
        return result;
    }

    public boolean isBalanceEnough() {
        for (String s : balanceInfo.keySet()) {
            if (s.startsWith("MINUS_")) {
                return false;
            }
        }
        return true;
    }

    private boolean needApprove() {
        for (String s : balanceInfo.keySet()) {
            if (s.startsWith("GAS_")) {
                return true;
            }
        }
        return false;
    }

    private Observable<String> approve() {
        Observable<String> result = Observable.just("failed");
        for (Map.Entry<String, Double> entry : balanceInfo.entrySet()) {
            if (entry.getKey().startsWith("GAS_")) {
                if (entry.getValue() != 1 && entry.getValue() != 2) {
                    continue;
                }
                String token = entry.getKey().split("_")[1];
                if (entry.getValue() == 1) {
                    result = approveOnce(token);
                } else if (entry.getValue() == 2) {
                    result = approveTwice(token);
                }
            }
        }
        return result;
    }

    protected Observable<String> approve(String symbol, Double value) {
        Transfer transfer = new Transfer(credentials);
        String contract = token.getTokenBySymbol(symbol).getProtocol();
        BigInteger amount = token.getWeiFromDouble(symbol, value);
        BigInteger gasPrice = gas.getCustomizeGasPriceInWei().toBigInteger();
        BigInteger gasLimit = gas.getGasLimitByType("approve");
        return transfer.erc20(contract, gasPrice, gasLimit)
                .approve(credentials, contract, Default.DELEGATE_ADDRESS, amount);
    }

    private Observable<String> approveOnce(String symbol) {
        Transfer transfer = new Transfer(credentials);
        String contract = token.getTokenBySymbol(symbol).getProtocol();
        BigInteger value = token.getWeiFromDouble(symbol, (double) Integer.MAX_VALUE);
        BigInteger gasPrice = gas.getCustomizeGasPriceInWei().toBigInteger();
        BigInteger gasLimit = gas.getGasLimitByType("approve");
        return transfer.erc20(contract, gasPrice, gasLimit)
                .approve(credentials, contract, Default.DELEGATE_ADDRESS, value);
    }

    private Observable<String> approveTwice(String symbol) {
        Transfer transfer = new Transfer(credentials);
        String contract = token.getTokenBySymbol(symbol).getProtocol();
        BigInteger value = BigInteger.ZERO;
        BigInteger gasPrice = gas.getCustomizeGasPriceInWei().toBigInteger();
        BigInteger gasLimit = gas.getGasLimitByType("approve");
        final Erc20TransactionManager manager = transfer.erc20(contract, gasPrice, gasLimit);
        return manager.approve(credentials, contract, Default.DELEGATE_ADDRESS, value)
                .observeOn(Schedulers.io())
                .flatMap((Func1<String, Observable<String>>) s -> {
                    BigInteger value1 = token.getWeiFromDouble(symbol, (double) Integer.MAX_VALUE);
                    return manager.approve(credentials, contract, Default.DELEGATE_ADDRESS, value1);
                });
    }

    protected abstract Observable<RelayResponseWrapper> submit();
}

File: loopr-ios/DataManagers/OrderDataManager.swift
//
//  OrderDataManager.swift
//  loopr-ios
//
//  Created by xiaoruby on 2/5/18.
//  Copyright © 2018 Loopring. All rights reserved.
//

import Foundation
import BigInt
import web3

// It's to all orders of an address.
class OrderDataManager {

    static let shared = OrderDataManager()
    let Eip191Header = "\u{0019}\u{0001}"
    let Eip712DomainHash = "0xaea25658c273c666156bd427f83a666135fcde6887a6c25fc1cd1562bc4f3f34"
    let Eip712OrderSchemaHash = "0x40b942178d2a51f1f61934268590778feb8114db632db7d88537c98d2b05c5f2"

    var errorInfo: [String: Any] = [:]
    var balanceInfo: [String: Double] = [:]

    let gasManager = GasDataManager.shared
    let tokenManager = TokenDataManager.shared
    let walletManager = CurrentAppWalletDataManager.shared
    let sendManager = SendCurrentAppWalletDataManager.shared

    // Similar naming in Trade.swift
    var baseToken: String = "LRC"
    var quoteToken: String = "WETH"
    var market: Market?

    func new(baseToken: String, quoteToken: String, market: Market) {
        self.baseToken = baseToken
        self.quoteToken = quoteToken
        self.market = market
    }

    var orders: [RawOrder] = []

    init() {
        orders = []
    }

    func getOrders(orderStatuses: [OrderStatus]? = nil) -> [RawOrder] {
        guard let orderStatuses = orderStatuses else {
            return orders
        }
        let filteredOrder = orders.filter { (order) -> Bool in
            orderStatuses.contains(order.state.status)
        }
        return filteredOrder
    }

    func getOrders(token: String? = nil) -> [RawOrder] {
        guard let token = token else {
            return orders
        }
        return orders.filter { (order) -> Bool in
            return order.tokenSell == token || order.tokenBuy == token
        }
    }

    func getOrdersFromServer(cursor: UInt, size: UInt = 20, statuses: [OrderStatus]? = nil, completionHandler: @escaping (_ error: Error?) -> Void) {
        if let owner = CurrentAppWalletDataManager.shared.getCurrentAppWallet()?.address {
            LoopringAPIRequest.getOrders(owner: owner, statuses: statuses, cursor: cursor, size: size) { result, error in
                guard let orders = result?.orders, error == nil else {
                    self.orders = []
                    completionHandler(error)
                    return
                }
                if cursor == 1 {
                    self.orders = orders
                } else {
                    self.orders += orders
                }
                completionHandler(error)
            }
        }
    }

    func constructOrder(side: OrderSide, amountBuy: Double, amountSell: Double, validSince: Int, validUntil: Int) -> RawOrder? {

        var result: RawOrder?
        var tokenB, tokenBuy, tokenS, tokenSell, amountB, amountS: String
        let address = CurrentAppWalletDataManager.shared.getCurrentAppWallet()!.address

        if let baseToken = tokenManager.getTokenBySymbol(baseToken),
           let quoteToken = tokenManager.getTokenBySymbol(quoteToken) {
            if side == .buy {
                tokenBuy = baseToken.symbol; tokenB = baseToken.address
                tokenSell = quoteToken.symbol; tokenS = quoteToken.address
                amountB = BigInt.generate(from: amountBuy, by: baseToken.decimals).toHexWithPrefix()
                amountS = BigInt.generate(from: amountSell, by: quoteToken.decimals).toHexWithPrefix()
            } else {
                tokenBuy = quoteToken.symbol; tokenB = quoteToken.address
                tokenSell = baseToken.symbol; tokenS = baseToken.address
                amountB = BigInt.generate(from: amountBuy, by: quoteToken.decimals).toHexWithPrefix()
                amountS = BigInt.generate(from: amountSell, by: baseToken.decimals).toHexWithPrefix()
            }

            let orderParams = OrderParams()
            orderParams.validUntil = validUntil
            let erc1400Params = ERC1400Params()
            let feeParams = FeeParams()
            let lrcFee = getLrcFee(amountSell, tokenSell)
            feeParams.tokenF = "LRC"
            feeParams.tokenFee = tokenManager.getAddress(by: "LRC")!
            feeParams.amountF = lrcFee
            feeParams.amountFee = BigInt.generate(from: lrcFee, by: 18).toHexWithPrefix()
            feeParams.tokenRecipient = address

            var order = RawOrder()
            order.owner = address; order.version = 0
            order.tokenB = tokenB; order.tokenBuy = tokenBuy
            order.tokenS = tokenS; order.tokenSell = tokenSell
            order.amountBuy = amountBuy; order.amountB = amountB
            order.amountSell = amountSell; order.amountS = amountS
            order.validSince = validSince
            order.params = orderParams
            order.feeParams = feeParams
            order.erc1400Params = erc1400Params

            result = completeOrder(&order)
        }
        return result
    }

    func completeOrder(_ order: inout RawOrder) -> RawOrder? {
        let orderData = getOrderHash(order: order)
        SendCurrentAppWalletDataManager.shared._keystore()
        if case (let signature?, let hash?) = web3swift.sign(message: orderData) {
            order.hash = hash
            order.r = signature.r
            order.s = signature.s
            order.v = UInt(signature.v)!
        }
    }

    func getOrderHash(order: RawOrder) -> Data {
        var bitStream: BitStream = BitStream()
        let orderParams = order.params
        let feeParams = order.feeParams
        let erc1400Params = order.erc1400Params

        let transferDataHash = erc1400Params.transferDataS.sha3(.keccak256).hexString
        try! bitStream.addBytes32(Eip712OrderSchemaHash, forceAppend: true)
        bitStream.addUint(order.amountS.toBigInt!, forceAppend: true);
        bitStream.addUint(order.amountB.toBigInt!, forceAppend: true);
        bitStream.addUint(feeParams.amountFee.toBigInt!, forceAppend: true);
        bitStream.addUint(BigInt(order.validSince), forceAppend: true);
        bitStream.addUint(BigInt(order.params.validUntil), forceAppend: true);
        bitStream.addAddress(order.owner, 32, forceAppend: true);
        bitStream.addAddress(order.tokenS, 32, forceAppend: true);
        bitStream.addAddress(order.tokenB, 32, forceAppend: true);
        bitStream.addAddress(orderParams.dualAuthAddr, 32, forceAppend: true);
        bitStream.addAddress(orderParams.broker, 32, forceAppend: true);
        bitStream.addAddress(orderParams.orderInterceptor, 32, forceAppend: true);
        bitStream.addAddress(orderParams.wallet, 32, forceAppend: true);
        bitStream.addAddress(feeParams.tokenRecipient, 32, forceAppend: true);
        bitStream.addAddress(feeParams.tokenFee, forceAppend: true);
        bitStream.addUint(BigInt(feeParams.walletSplitPercentage), forceAppend: true);
        bitStream.addUint(BigInt(feeParams.tokenSFeePercentage), forceAppend: true);
        bitStream.addUint(BigInt(feeParams.tokenBFeePercentage), forceAppend: true);
        bitStream.addBool(orderParams.allOrNone, forceAppend: true);
        bitStream.addUint(BigInt(erc1400Params.tokenStandardS), forceAppend: true);
        bitStream.addUint(BigInt(erc1400Params.tokenStandardB), forceAppend: true);
        bitStream.addUint(BigInt(erc1400Params.tokenStandardFee), forceAppend: true);
        try! bitStream.addBytes32(erc1400Params.trancheS, forceAppend: true);
        try! bitStream.addBytes32(erc1400Params.trancheB, forceAppend: true);
        try! bitStream.addBytes32(transferDataHash, forceAppend: true);
        
        let orderDataHash: String = bitStream.getBytes().sha3(.keccak256).hexString
        let outerStream: BitStream = BitStream()
        outerStream.addHex(Eip191Header.hexString, forceAppend: true)
        try! outerStream.addBytes32(Eip712DomainHash, forceAppend: true)
        try! outerStream.addBytes32(orderDataHash, forceAppend: true)
        return outerStream.getBytes().sha3(.keccak256).hexString
    }

    func getLrcFee(_ amountS: Double, _ tokenS: String) -> Double {
        var result: Double = 0
        let ratio = SettingDataManager.shared.getLrcFeeRatio()
        if let price = PriceDataManager.shared.getPrice(of: tokenS),
           let lrcPrice = PriceDataManager.shared.getPrice(of: "LRC") {
            result = price * amountS * ratio / lrcPrice
        }
        let minLrc = GasDataManager.shared.getGasAmount(by: "eth_transfer", in: "LRC")
        return max(result, minLrc)
    }
}
